### PR TITLE
Improve SQL Server type mapping documentation

### LINKED ISSUE #11280: Add more mapping relationships to the SQL Server type mapping doc
https://github.com/trinodb/trino/pull/11274#issuecomment-1056863517

### PR DESCRIPTION
**Pull request by tangjiangling**
## Description

> Is this change a fix, improvement, new feature, refactoring, or other?

Improvement.

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)

SQL Server type mapping doc.

> How would you describe this change to a non-technical end user or system administrator?

N/A.

## Related issues, pull requests, and links

Fixes #11280
Related #11448

## Documentation

( ) No documentation is needed.
(x) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

(x) No release notes entries required.
( ) Release notes entries required with the following suggested text:

### CODE REVIEW

**Comment by tangjiangling** [comment]
Note: We mapped Trino `CHAR/VARCHAR` to SQL Server `nchar/nvarchar` in write operation, and the minimum length of SQL Server `nchar/nvarchar` is 1. And In the type mapping test we **did not** test what happens when we write `CHAR(0)/VARCHAR(0)` to SQL Server.

I'll send a pr to fix this.

File: docs/src/main/sphinx/connector/sqlserver.rst [hunk 1]
```diff
@@ -119,26 +119,186 @@ that catalog name instead of ``sqlserver`` in the above examples.
...
-``char(n)``                         ``char(n)``
-``varchar(n)``                      ``varchar(n)``
-``date``                            ``date``
-``datetime2(n)``                    ``timestamp(n)``
-``datetimeoffset(n)``               ``timestamp(n) with time zone``
-==================================  ===============================
+Because Trino and SQL Server each support types that the other does not, this
+connector modifies some types when reading or writing data.
+
+SQL Server to Trino read type mapping
```
Comment (jhlodin) [inline docs/src/main/sphinx/connector/sqlserver.rst:125]: Is "read type mapping" the best term to use? That would imply that this is only relevant when performing read operations (https://trino.io/docs/current/language/sql-support.html#read-operations), but I'd suggest it's better to just call this "SQL Server type to Trino type mapping", and the following section "Trino type to SQL Server type mapping" without read/write being part of it

File: docs/src/main/sphinx/connector/sqlserver.rst [hunk 2]
```diff
@@ -119,26 +119,186 @@ that catalog name instead of ``sqlserver`` in the above examples.
...
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+This connector supports reading the following SQL Server types and performs
+conversion to Trino types with the detailed mappings as shown in the following
+table.
```
Comment (jhlodin) [inline docs/src/main/sphinx/connector/sqlserver.rst:130]: See comment above. Would suggest "The connector maps SQL server types to the corresponding Trino types following this table:"

File: docs/src/main/sphinx/connector/sqlserver.rst [hunk 3]
```diff
@@ -119,26 +119,186 @@ that catalog name instead of ``sqlserver`` in the above examples.
...
+  * - ``DATETIMEOFFSET[(n)]``
+    - ``TIMESTAMP(n) WITH TIME ZONE``
+    - ``0 <= n <= 7``
+
+Trino to SQL Server write type mapping
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+This connector supports writing the following Trino types and performs
+conversion to SQL Server types with the detailed mappings as shown in the
+following table.
```
Comment (jhlodin) [inline docs/src/main/sphinx/connector/sqlserver.rst:220]: See comment above. Would suggest rewording to "The connector maps Trino to the corresponding SQL Server types following this table:"

File: docs/src/main/sphinx/connector/sqlserver.rst [hunk 4]
```diff
@@ -119,26 +119,186 @@ that catalog name instead of ``sqlserver`` in the above examples.
...
+    - ``TIMESTAMP(n)``
+    - ``0 <= n <= 7``
+  * - ``SMALLDATETIME``
+    - ``TIMESTAMP(0)``
+    -
+  * - ``DATETIMEOFFSET[(n)]``
+    - ``TIMESTAMP(n) WITH TIME ZONE``
+    - ``0 <= n <= 7``
+
+Trino to SQL Server write type mapping
```
Comment (jhlodin) [inline docs/src/main/sphinx/connector/sqlserver.rst:215]: See comment above, would suggest "SQL Server type to Trino type mapping"

File: docs/src/main/sphinx/connector/sqlserver.rst [hunk 5]
```diff
@@ -119,26 +119,186 @@ that catalog name instead of ``sqlserver`` in the above examples.
...
-``int``                             ``integer``
-``float``                           ``double``
-``char(n)``                         ``char(n)``
-``varchar(n)``                      ``varchar(n)``
-``date``                            ``date``
-``datetime2(n)``                    ``timestamp(n)``
-``datetimeoffset(n)``               ``timestamp(n) with time zone``
-==================================  ===============================
+Because Trino and SQL Server each support types that the other does not, this
+connector modifies some types when reading or writing data.
```
Comment (jhlodin) [inline docs/src/main/sphinx/connector/sqlserver.rst:123]: Would suggest adding a sentence clarifying that the type mappings are not bi-directional. Something like "Data types may not map the same way in both directions between Trino and the data source. Refer to the following sections for type mapping in each direction." or similar

File: docs/src/main/sphinx/connector/sqlserver.rst [hunk 6]
```diff
@@ -119,26 +119,186 @@ that catalog name instead of ``sqlserver`` in the above examples.
...
+  * - ``TIMESTAMP(n)``
+    - ``DATETIME2(n)``
+    - ``0 <= n <= 7``
 
 Complete list of `SQL Server data types
 <https://msdn.microsoft.com/en-us/library/ms187752.aspx>`_.
 
+.. _sqlserver-numeric-mapping:
+
+Mapping numeric types
```
Comment (jhlodin) [inline docs/src/main/sphinx/connector/sqlserver.rst:277]: Numeric type mapping

File: docs/src/main/sphinx/connector/sqlserver.rst [hunk 7]
```diff
@@ -119,26 +119,186 @@ that catalog name instead of ``sqlserver`` in the above examples.
...
+
+For SQL Server ``FLOAT[(n)]``:
+
+- If ``n`` is not specified maps to Trino ``Double``
+- If ``1 <= n <= 24`` maps to Trino ``REAL``
+- If ``24 < n <= 53`` maps to Trino ``DOUBLE``
+
+.. _sqlserver-character-mapping:
+
+Mapping character types
```
Comment (jhlodin) [inline docs/src/main/sphinx/connector/sqlserver.rst:288]: Character type mapping

**Review by jhlodin — COMMENTED** [review]
Thanks for proposing these changes! Definitely an improvement. I commented some proposed changes that may be worth discussing, just want to find the best possible wording so we can start applying these changes to other connectors as well.

File: docs/src/main/sphinx/connector/sqlserver.rst (hunk 5 above)
Comment (tangjiangling) [reply docs/src/main/sphinx/connector/sqlserver.rst:123]: Good suggestion!

File: docs/src/main/sphinx/connector/sqlserver.rst (hunk 1 above)
Comment (tangjiangling) [reply docs/src/main/sphinx/connector/sqlserver.rst:125]: Good suggestion!

**Comment by tangjiangling** [comment]
@jhlodin Thanks for the review, it was very useful ~
And I will update this pr later today.

**Comment by tangjiangling** [comment]
@ebyhr @hashhar Please also take a look to see if the mapping is correct.

**Review by ebyhr — COMMENTED** [review]
Could you remove types where we don't have tests? e.g. XML, uniqueidentifier and so on

**Comment by tangjiangling** [comment]
> Could you remove types where we don't have tests? e.g. XML, uniqueidentifier and so on

Removed `MONEY`, `SMALLMONEY`, `uniqueidentifier`, `XML`, `BINARY`, ``IMAGE``, `TIMESTAMP`, `UDT`, `GEOMETRY`, `GEOGRAPHY`.

If there is a strong demand from users one day, we can add it in.

**Review by mosabua — COMMENTED** [review]
This is a great improvement. In the interest of making this a model for all other connectors I would like some input from @hashhar @electrum and others about better titles and descriptive text for the two mappings. 

Here is my current (maybe faulty undestanding).

One mapping is for creating new columns in CTAS, ALTER TABLE, CREATE TABLE.

The other mapping is for working with existing columns. And that means for read (SELECT), but also write operations such as INSERT and UPDATE.. 

If this is correct .. what could be good titles for the two sections of mappings. 

DB to Trino

and 

Trino to DB 

as we use them now are anything but intuitive ... we should come up with something better.. but what? Any suggestions?

**Comment by tangjiangling** [comment]
Ping, could anyone help to review it? (so I can move forward with fixing #11448)

**Comment by jhlodin** [comment]
> Ping, could anyone help to review it? (so I can move forward with fixing #11448)

 +1, would be great to use this as a jumping-off point for improving the rest of our connectors' type mapping sections.
@hashhar @electrum

**Review by electrum — APPROVED** [review]

**Comment by electrum** [comment]
Thanks! This is a big improvement.
